src/librustc_mir/const_eval/mod.rs

```diff
@@ -76,9 +76,16 @@ pub(crate) fn deref_const<'tcx>(
     val: &'tcx ty::Const<'tcx>,
 ) -> &'tcx ty::Const<'tcx> {
     trace!("deref_const: {:?}", val);
-    let ecx = mk_eval_cx(tcx, DUMMY_SP, param_env, false);
+    let mut ecx = mk_eval_cx(tcx, DUMMY_SP, param_env, false);
     let op = ecx.const_to_op(val, None).unwrap();
     let mplace = ecx.deref_operand(op).unwrap();
+    if let Scalar::Ptr(ptr) = mplace.ptr {
+        assert!(
+            ecx.memory.get_raw_mut(ptr.alloc_id).is_err(),
```

Comment:
That's not the right test I think -- this will err even for mutable statics as those may not be mutated by CTFE.
You'd have to check the allocation [`mutability` field](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/interpret/allocation/struct.Allocation.html#structfield.mutability).

However, isn't it already the case that when pattern matching reads from an allocation, it uses the CTFE machine, which refuses to read from mutable memory? Not sure if there's a way to test this, but if that is true, no further check would be needed here.

Reply:
> it uses the CTFE machine, which refuses to read from mutable memory?

It refuses to read from mutable statics, but we don't actually check against reading from allocations marked as mutable afaict.
I don't think we can test this, even with unleash mode, as this is not part of const eval really, it's happening later than that.

Reply:
We have this:

https://github.com/rust-lang/rust/blob/1a22a0ff93d63f738151f096434e732466b4a42e/src/librustc_mir/const_eval/machine.rs#L363

and that's called for all accesses to global memory, not just statics:

https://github.com/rust-lang/rust/blob/1a22a0ff93d63f738151f096434e732466b4a42e/src/librustc_mir/interpret/memory.rs#L486

Reply:
But it might be worth adding a comment at that `before_access_global` call site that pattern matching soundness relies on this being done for all accesses.

Reply:
I moved to checking the `mutability` field

Reply:
Looks good. :)